File: src/dartlab/viz/display/finance/accounts.py
"""K-IFRS 표준 항목 28 종 + 시리즈 추출.

회사마다 account_nm 이 다름 (카카오 "영업수익" vs 삼성 "매출액").
IFRS taxonomy account_id 우선 + account_nm 키워드 fallback.

Returns 형식 통일: extractSeries 는 list[float|None] 반환 (periods 길이 일치).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

import polars as pl


@dataclass(frozen=True)
class StandardAccount:
    """표준 재무 항목."""

    key: str
    label: str
    sjDiv: str  # "IS" | "BS" | "CF"
    ifrsIds: tuple[str, ...]
    nameKeywords: tuple[str, ...]


_STANDARDS: tuple[StandardAccount, ...] = (
    # ─── IS 10 ───
    StandardAccount(
        "revenue", "매출액", "IS", ("ifrs-full_Revenue", "ifrs_Revenue"), ("매출액", "영업수익", "수익(매출액)", "매출")
    ),
    StandardAccount("costOfSales", "매출원가", "IS", ("ifrs-full_CostOfSales", "ifrs_CostOfSales"), ("매출원가",)),
    StandardAccount("grossProfit", "매출총이익", "IS", (), ("매출총이익",)),
    StandardAccount(
        "operatingIncome",
        "영업이익",
        "IS",
        ("dart_OperatingIncomeLoss", "ifrs-full_ProfitLossFromOperatingActivities", "ifrs_OperatingProfitLoss"),
        ("영업이익", "영업이익(손실)"),
    ),
    StandardAccount(
        "netIncome",
        "당기순이익",
        "IS",
        ("ifrs-full_ProfitLoss", "ifrs_ProfitLoss"),
        ("당기순이익", "당기순이익(손실)", "순이익"),
    ),
    StandardAccount(
        "sga",
        "판매비와관리비",
        "IS",
        ("dart_TotalSellingGeneralAdministrativeExpenses", "ifrs-full_SellingGeneralAndAdministrativeExpense"),
        ("판매비와관리비", "판매관리비"),
    ),
    StandardAccount("rnd", "연구개발비", "IS", ("dart_ResearchAndDevelopmentExpenses",), ("연구개발",)),
    StandardAccount(
        "financeIncome", "금융수익", "IS", ("ifrs-full_FinanceIncome", "ifrs_FinanceIncome"), ("금융수익",)
    ),
    StandardAccount("financeCosts", "금융비용", "IS", ("ifrs-full_FinanceCosts", "ifrs_FinanceCosts"), ("금융비용",)),
    StandardAccount(
        "incomeTax",
        "법인세비용",
        "IS",
        ("ifrs-full_IncomeTaxExpenseContinuingOperations", "ifrs_IncomeTaxExpense"),
        ("법인세비용",),
    ),
    # ─── BS 14 ───
    StandardAccount("assets", "자산총계", "BS", ("ifrs-full_Assets", "ifrs_Assets"), ("자산총계",)),
    StandardAccount(
        "currentAssets", "유동자산", "BS", ("ifrs-full_CurrentAssets", "ifrs_CurrentAssets"), ("유동자산",)
    ),
    StandardAccount(
        "nonCurrentAssets", "비유동자산", "BS", ("ifrs-full_NoncurrentAssets", "ifrs_NoncurrentAssets"), ("비유동자산",)
    ),
    StandardAccount(
        "cash",
        "현금및현금성자산",
        "BS",
        ("ifrs-full_CashAndCashEquivalents", "ifrs_CashAndCashEquivalents"),
        ("현금및현금성자산", "현금성자산"),
    ),
    StandardAccount("inventories", "재고자산", "BS", ("ifrs-full_Inventories", "ifrs_Inventories"), ("재고자산",)),
    StandardAccount(
        "receivables",
        "매출채권",
        "BS",
        ("ifrs-full_TradeAndOtherCurrentReceivables", "dart_ShortTermTradeReceivable"),
        ("매출채권",),
    ),
    StandardAccount("liabilities", "부채총계", "BS", ("ifrs-full_Liabilities", "ifrs_Liabilities"), ("부채총계",)),
    StandardAccount(
        "currentLiabilities",
        "유동부채",
        "BS",
        ("ifrs-full_CurrentLiabilities", "ifrs_CurrentLiabilities"),
        ("유동부채",),
    ),
    StandardAccount(
        "nonCurrentLiabilities",
        "비유동부채",
        "BS",
        ("ifrs-full_NoncurrentLiabilities", "ifrs_NoncurrentLiabilities"),
        ("비유동부채",),
    ),
    StandardAccount(
        "payables",
        "매입채무",
        "BS",
        ("ifrs-full_TradeAndOtherCurrentPayables", "dart_ShortTermTradePayables"),
        ("매입채무",),
    ),
    StandardAccount(
        "shortDebt", "단기차입금", "BS", ("ifrs-full_ShorttermBorrowings", "ifrs_ShorttermBorrowings"), ("단기차입금",)
    ),
    StandardAccount(
        "longDebt",
        "장기차입금",
        "BS",
        ("ifrs-full_LongtermBorrowings", "ifrs_LongtermBorrowings"),
        ("장기차입금", "사채"),
    ),
    StandardAccount("equity", "자본총계", "BS", ("ifrs-full_Equity", "ifrs_Equity"), ("자본총계",)),
    StandardAccount(
        "retainedEarnings", "이익잉여금", "BS", ("ifrs-full_RetainedEarnings", "ifrs_RetainedEarnings"), ("이익잉여금",)
    ),
    # ─── CF 4 ───
    StandardAccount(
        "cfOperating",
        "영업활동현금흐름",
        "CF",
        ("ifrs-full_CashFlowsFromUsedInOperatingActivities", "ifrs_CashFlowsFromUsedInOperatingActivities"),
        ("영업활동", "영업활동현금흐름"),
    ),
    StandardAccount(
        "cfInvesting",
        "투자활동현금흐름",
        "CF",
        ("ifrs-full_CashFlowsFromUsedInInvestingActivities", "ifrs_CashFlowsFromUsedInInvestingActivities"),
        ("투자활동", "투자활동현금흐름"),
    ),
    StandardAccount(
        "cfFinancing",
        "재무활동현금흐름",
        "CF",
        ("ifrs-full_CashFlowsFromUsedInFinancingActivities", "ifrs_CashFlowsFromUsedInFinancingActivities"),
        ("재무활동", "재무활동현금흐름"),
    ),
    StandardAccount(
        "capex",
        "유형자산취득",
        "CF",
        (
            "ifrs-full_PurchaseOfPropertyPlantAndEquipmentClassifiedAsInvestingActivities",
            "dart_PurchaseOfPropertyPlantAndEquipment",
        ),
        ("유형자산의취득", "유형자산취득"),
    ),
    StandardAccount(
        "dividendsPaid",
        "배당금지급",
        "CF",
        ("ifrs-full_DividendsPaidClassifiedAsFinancingActivities", "ifrs_DividendsPaid"),
        ("배당금지급",),
    ),
)

_BY_KEY: dict[str, StandardAccount] = {s.key: s for s in _STANDARDS}


def listStandards(sjDiv: str | None = None) -> list[StandardAccount]:
    """표준 항목 카탈로그. sjDiv (BS/IS/CF) 지정 시 해당 영역만."""
    if sjDiv is None:
        return list(_STANDARDS)
    return [s for s in _STANDARDS if s.sjDiv == sjDiv]


def standard(key: str) -> StandardAccount:
    """key 로 표준 항목 1 개 조회."""
    return _BY_KEY[key]


def extractSeries(
    norm: pl.DataFrame,
    key: str,
    periods: Iterable[str],
    *,
    fsDiv: str = "CFS",
) -> list[float | None]:
    """표준 항목 N 기간 값 추출 (list 반환, periods 길이 일치).

    fsDiv: CFS (연결) 우선, 없으면 OFS (별도) fallback.
    accountId IFRS 매칭 우선, 안되면 accountNm 키워드 매칭.
    """
    periods_list = list(periods)
    if norm is None or norm.height == 0:
        return [None] * len(periods_list)
    sa = _BY_KEY.get(key)
    if sa is None:
        return [None] * len(periods_list)

    sub = norm.filter(
        (pl.col("sjDiv") == sa.sjDiv) & (pl.col("fsDiv") == fsDiv) & (pl.col("period").is_in(periods_list))
    )
    if sub.height == 0 and fsDiv == "CFS":
        sub = norm.filter(
            (pl.col("sjDiv") == sa.sjDiv) & (pl.col("fsDiv") == "OFS") & (pl.col("period").is_in(periods_list))
        )
    if sub.height == 0:
        return [None] * len(periods_list)

    idCond = pl.col("accountId").is_in(list(sa.ifrsIds)) if sa.ifrsIds else pl.lit(False)
    nmCond = pl.lit(False)
    for kw in sa.nameKeywords:
        nmCond = nmCond | (pl.col("accountNm").str.strip_chars() == kw)
    cond = idCond | nmCond
    matched = sub.filter(cond)
    if matched.height == 0:
        return [None] * len(periods_list)

    matched = matched.with_columns(pl.when(idCond).then(2).when(nmCond).then(1).otherwise(0).alias("_priority"))
    agg = (
        matched.sort(["_priority", "ord"], descending=[True, False], nulls_last=True)
        .group_by("period")
        .agg(pl.col("amount").first().alias("amount"))
    )
    lookup = dict(zip(agg["period"].to_list(), agg["amount"].to_list(), strict=True))
    return [lookup.get(p) for p in periods_list]

File: ui/packages/runtime/src/data/finance/accounts.ts
// 재무제표 28 표준계정 매핑 + 계정매칭 — 단일 SSOT.
//
// 원래 financeSource.ts(터미널 번들)에 인라인돼 있던 STD 테이블·매칭 규칙·파싱 primitive 를
// 여기로 추출한다. financeSource.ts(브라우저 터미널)와 annual.ts(블로그·정적 SEO 빌드타임)가
// **같은 표준화**를 공유 — 평행 재구현 추가 0. src/dartlab/viz/display/finance/accounts.py(_STANDARDS)
// 포팅의 TS 정본.
//
// Node+브라우저 안전(svelte·DOM 의존 0) — annual.ts 가 SvelteKit prerender(Node)에서 import 한다.

export interface StdAcct {
	key: string;
	sj: 'IS' | 'BS' | 'CF' | 'CIS';
	ids: string[]; // account_id (IFRS) 우선
	kw: string[]; // account_nm 키워드 fallback
	ex?: string[]; // nm 매칭 제외 키워드 — includes 포함매칭의 오선택 차단 (예: longDebt 가 '유동성장기차입금'을 잡는 사고)
}

// ── 28 표준계정 (accounts.py _STANDARDS 포팅) ──
export const STD: StdAcct[] = [
	// IS
	{ key: 'revenue', sj: 'IS', ids: ['ifrs-full_Revenue', 'ifrs_Revenue'], kw: ['매출액', '영업수익', '수익(매출액)'] },
	{ key: 'costOfSales', sj: 'IS', ids: ['ifrs-full_CostOfSales', 'ifrs_CostOfSales'], kw: ['매출원가'] },
	{ key: 'grossProfit', sj: 'IS', ids: [], kw: ['매출총이익'] },
	{ key: 'operatingIncome', sj: 'IS', ids: ['dart_OperatingIncomeLoss', 'ifrs-full_ProfitLossFromOperatingActivities', 'ifrs_OperatingProfitLoss'], kw: ['영업이익', '영업이익(손실)'] },
	{ key: 'netIncome', sj: 'IS', ids: ['ifrs-full_ProfitLoss', 'ifrs_ProfitLoss'], kw: ['당기순이익', '당기순이익(손실)', '순이익'] },
	{ key: 'sga', sj: 'IS', ids: ['dart_TotalSellingGeneralAdministrativeExpenses', 'ifrs-full_SellingGeneralAndAdministrativeExpense'], kw: ['판매비와관리비', '판매관리비'] },
	{ key: 'financeIncome', sj: 'IS', ids: ['ifrs-full_FinanceIncome', 'ifrs_FinanceIncome'], kw: ['금융수익'] },
	{ key: 'financeCosts', sj: 'IS', ids: ['ifrs-full_FinanceCosts', 'ifrs_FinanceCosts'], kw: ['금융비용'] },
	{ key: 'incomeTax', sj: 'IS', ids: ['ifrs-full_IncomeTaxExpenseContinuingOperations', 'ifrs_IncomeTaxExpense'], kw: ['법인세비용'] },
	// BS
	{ key: 'assets', sj: 'BS', ids: ['ifrs-full_Assets', 'ifrs_Assets'], kw: ['자산총계'] },
	{ key: 'currentAssets', sj: 'BS', ids: ['ifrs-full_CurrentAssets', 'ifrs_CurrentAssets'], kw: ['유동자산'] },
	{ key: 'cash', sj: 'BS', ids: ['ifrs-full_CashAndCashEquivalents', 'ifrs_CashAndCashEquivalents'], kw: ['현금및현금성자산', '현금성자산'] },
	{ key: 'inventories', sj: 'BS', ids: ['ifrs-full_Inventories', 'ifrs_Inventories'], kw: ['재고자산'] },
	{ key: 'receivables', sj: 'BS', ids: ['ifrs-full_TradeAndOtherCurrentReceivables', 'dart_ShortTermTradeReceivable'], kw: ['매출채권'] },
	{ key: 'liabilities', sj: 'BS', ids: ['ifrs-full_Liabilities', 'ifrs_Liabilities'], kw: ['부채총계'] },
	{ key: 'currentLiabilities', sj: 'BS', ids: ['ifrs-full_CurrentLiabilities', 'ifrs_CurrentLiabilities'], kw: ['유동부채'] },
	{ key: 'payables', sj: 'BS', ids: ['ifrs-full_TradeAndOtherCurrentPayables', 'dart_ShortTermTradePayables'], kw: ['매입채무'] },
	{ key: 'shortDebt', sj: 'BS', ids: ['ifrs-full_ShorttermBorrowings', 'ifrs_ShorttermBorrowings'], kw: ['단기차입금'], ex: ['유동성'] },
	// '유동성장기차입금'.includes('장기차입금')=true 라 ex 가드 없이는 longDebt 가 유동성 행을
	// 오선택/이중계상 — 차입 3분해(shortDebt·currentLtDebt·longDebt)의 정합 전제.
	{ key: 'longDebt', sj: 'BS', ids: ['ifrs-full_LongtermBorrowings', 'ifrs_LongtermBorrowings'], kw: ['장기차입금', '사채'], ex: ['유동성'] },
	{ key: 'currentLtDebt', sj: 'BS', ids: [], kw: ['유동성장기차입금', '유동성장기부채', '유동성사채'] },
	{ key: 'equity', sj: 'BS', ids: ['ifrs-full_Equity', 'ifrs_Equity'], kw: ['자본총계'] },
	{ key: 'capitalStock', sj: 'BS', ids: ['ifrs-full_IssuedCapital', 'ifrs_IssuedCapital', 'dart_IssuedCapital'], kw: ['자본금'], ex: ['잉여금'] },
	{ key: 'capitalSurplus', sj: 'BS', ids: ['ifrs-full_SharePremium', 'dart_AdditionalPaidInCapital', 'ifrs_SharePremium'], kw: ['자본잉여금', '주식발행초과금'] },
	{ key: 'retainedEarnings', sj: 'BS', ids: ['ifrs-full_RetainedEarnings', 'ifrs_RetainedEarnings'], kw: ['이익잉여금'] },
	// CF
	{ key: 'cfOperating', sj: 'CF', ids: ['ifrs-full_CashFlowsFromUsedInOperatingActivities', 'ifrs_CashFlowsFromUsedInOperatingActivities'], kw: ['영업활동현금흐름', '영업활동'] },
	{ key: 'cfInvesting', sj: 'CF', ids: ['ifrs-full_CashFlowsFromUsedInInvestingActivities', 'ifrs_CashFlowsFromUsedInInvestingActivities'], kw: ['투자활동현금흐름', '투자활동'] },
	{ key: 'cfFinancing', sj: 'CF', ids: ['ifrs-full_CashFlowsFromUsedInFinancingActivities', 'ifrs_CashFlowsFromUsedInFinancingActivities'], kw: ['재무활동현금흐름', '재무활동'] },
	{ key: 'capex', sj: 'CF', ids: ['ifrs-full_PurchaseOfPropertyPlantAndEquipmentClassifiedAsInvestingActivities', 'dart_PurchaseOfPropertyPlantAndEquipment'], kw: ['유형자산의취득', '유형자산취득'] },
	{ key: 'dividendsPaid', sj: 'CF', ids: ['ifrs-full_DividendsPaidClassifiedAsFinancingActivities', 'ifrs_DividendsPaid'], kw: ['배당금지급'] },
	// CIS — 포괄손익계산서 원본 레인. 단일 포괄손익 회사(카카오류)의 손익 폴백·포괄손익 격차·SCE OCI 폴백.
	{ key: 'cisNetIncome', sj: 'CIS', ids: ['ifrs-full_ProfitLoss', 'ifrs_ProfitLoss'], kw: ['당기순이익', '분기순이익', '반기순이익'] },
	{ key: 'cisComprehensive', sj: 'CIS', ids: ['ifrs-full_ComprehensiveIncome', 'ifrs_ComprehensiveIncome'], kw: ['총포괄손익', '총포괄이익'] }
];

export const STD_BY_KEY: Record<string, StdAcct> = Object.fromEntries(STD.map((s) => [s.key, s]));
export const isStock = (k: string): boolean => STD_BY_KEY[k]?.sj === 'BS';

// thstrm_add_amount = 누적(YTD) — 정량재무 표(분기 standalone)가 쓴다. 표·차트 1회 다운로드 공유 위해 컬럼셋 통일.
export const FINANCE_COLUMNS = ['sj_div', 'fs_div', 'reprt_code', 'rcept_no', 'bsns_year', 'account_id', 'account_nm', 'account_detail', 'thstrm_amount', 'thstrm_add_amount', 'ord'];

export const Q_BY_CODE: Record<string, number> = { '11013': 1, '11012': 2, '11014': 3, '11011': 4 };

export interface RawRow extends Record<string, unknown> {
	sj_div?: string | null;
	fs_div?: string | null;
	reprt_code?: string | null;
	rcept_no?: string | null;
	bsns_year?: string | number | null;
	account_id?: string | null;
	account_nm?: string | null;
	account_detail?: string | null;
	thstrm_amount?: string | number | null;
	thstrm_add_amount?: string | number | null;
	ord?: string | number | null;
}

export interface Parsed {
	sj: string;
	year: number;
	q: number; // 1..4
	id: string;
	nm: string;
	detail: string;
	ord: number;
	amt: number; // 누적(YTD) for IS/CF, 시점 for BS
}

export function num(v: unknown): number | null {
	if (typeof v === 'number') return Number.isFinite(v) ? v : null;
	if (typeof v === 'bigint') return Number(v);
	if (typeof v === 'string' && v.trim()) {
		const n = Number(v.replace(/,/g, ''));
		return Number.isFinite(n) ? n : null;
	}
	return null;
}

// 표준계정 × (year,q) 격자 — STD 별 독립 매칭(행 소비/순서 의존 없음). id 매칭 우선, 없으면 nm 키워드.
// 같은 셀 다중 후보 시 account_detail='-' 우선·ord 최소. financeSource buildBundle 의 인라인 로직과 동일.
export function buildGrid(parsed: Parsed[]): Record<string, Map<string, Parsed>> {
	const score = (x: Parsed, byId: boolean) => (byId ? 0 : 1000) + (x.detail === '-' || x.detail === '' ? 0 : 100) + Math.min(x.ord, 99);
	const grid: Record<string, Map<string, Parsed>> = {};
	for (const s of STD) {
		const m = new Map<string, Parsed>();
		for (const p of parsed) {
			if (p.sj !== s.sj) continue;
			const idHit = s.ids.length > 0 && s.ids.includes(p.id);
			const nmHit = !idHit && s.kw.some((k) => p.nm.includes(k)) && !s.ex?.some((x) => p.nm.includes(x));
			if (!idHit && !nmHit) continue;
			const pk = `${p.year}-${p.q}`;
			const cur = m.get(pk);
			if (!cur || score(p, idHit) < score(cur, s.ids.includes(cur.id))) m.set(pk, p);
		}
		grid[s.key] = m;
	}
	return grid;
}
